---
title: "RNA-seq analysis in R"
subtitle: "Pre-processsing RNA-seq data"
author: "Stephane Ballereau, Chandra Chilamakuri, Dominique-Laurent Couturier, Mark Dunning, Abbi Edwards, Ashley Sawle"
date: '`r format(Sys.time(), "Last modified: %d %b %Y")`'
output:
  html_document:
    toc: yes
minutes: 300
layout: page
bibliography: ref.bib
editor_options: 
  chunk_output_type: inline
---

# Introduction

In this section we will begin the process of analysing the RNAseq data in R. In the
next section we will use [DESeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html)
for differential analysis. A detailed analysis workflow, recommended by the authors of DESeq2 can be found
on [the Bionconductor 
website](http://master.bioconductor.org/packages/release/workflows/vignettes/rnaseqGene/inst/doc/rnaseqGene.html).

We first need to:

* import our counts into R
* manipulate the imported data so that it is in the correct format for DESeq2
* filter out unwanted genes
* run some initial QC on the raw count data

# Data import

First, let's load all the packages we will need to analyse the data.

```{r setup, message = FALSE}
library(DESeq2)
library(tidyverse)
```

## Mouse mammary gland dataset

The data for this tutorial comes from a Nature Cell Biology paper,
[*EGF-mediated induction of Mcl-1 at the switch to lactation is essential for
alveolar cell survival*](https://www.nature.com/articles/ncb3117)
[@Fu2015]. The raw data (sequence reads) can be downloaded from the 'Sequence Read Archive' (SRA) under
[SRP045534](https://trace.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?study=SRP045534),
and processed data (counts) can be downloaded from Gene Expression Omnibus 
database (GEO) under accession number
[GSE60450](http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE60450). Please
see extended 
material for 
instructions on 
[downloading raw files from SRA](../Supplementary_Materials/S1_Getting_raw_reads_from_SRA.html) 
and [aligning fastq using HISAT2](../extended_html/C_Alignment_with_HISAT2_practical.html).

This study examines the expression profiles of basal stem-cell enriched cells
(B) and committed luminal cells (L) in the mammary gland of virgin, pregnant
and lactating mice. Six groups are present, with one for each combination of
cell type and mouse status. Each group contains two biological replicates.

## Reading in the sample metadata

The `SampleInfo.txt` file contains basic information about the samples that we will
need for the analysis today: name, cell type, status.

```{r loadSampleInfo, message = FALSE}
# Read the sample information into a data frame
sampleinfo <- read_tsv("data/SampleInfo.txt")
sampleinfo
```

## Reading in the count data

The raw reads were aligned using HISAT2 [@Kim2015] to the GRCm38 mouse reference
genome from Ensembl. featureCounts [@Liao2014] was used to count reads against
the Ensembl gene annotation and generate a counts matrix (as described in 
[the previous session](E_Read_Counts_with_Subread.practical.html)).

First we need to read the data into R from the file kept in the *data* directory.

```{r loadData, message = FALSE}
# Read the data into R
seqdata <- read_tsv("data/GSE60450_Lactation.featureCounts", comment="#")
seqdata
```

In the `seqdata` object each row represents a gene and columns are:

   1 *Geneid* - Gene Ensembl ID  
 2-5 *Chr*, *Start*, *End*, *Strand* - Genomic locations of exons of the gene  
   6 *Length* - Transcript length of the gene  
7-18 One column for each sample with the number of reads assigned to the gene
by featureCounts.

### A quick intro to `dplyr`

One of the most complex aspects of learning to work with data in `R` is 
getting to grips with subsetting and manipulating data tables. The package 
`dplyr` [@Wickham2018] was developed to make this process more intuitive than it
is using standard base `R` processes. It also makes use of a new symbol `%>%`,
called the "pipe", which makes the code a bit tidier. 

In particular we will use the commands:

* `select` to select columns from a table
* `filter` to filter rows based on the contents of a column in the table
* `rename` to rename columns

We will encounter a few more `dplyr` commands during the course, we will explain
their use as we come to them.

If you are familiar with R but not `dplyr` or `tidyverse` then we have a very
brief introduction [here](../extended_html/02a_A_brief_intro_to_dplyr.html). A more detailed 
introduction can be found in our [online R course](https://bioinformatics-core-shared-training.github.io/r-intro/week4.html)

## Format the data

We will be manipulating and reformating the counts matrix into a suitable
format for DESeq2.

The `seqdata` object is a `dataframe` in which the first six columns contain gene annotation
information and the remaining columns the count data.

DESeq2 requires a simple object containing only the count data, we'll keep the
gene ID by setting them as the row names.

We will create a new counts data object, `countdata`, that contains only the
counts for the 12 samples.  

Our `sampleinfo` object contains a column with the sample names. We should
adjust the column names of our count matrix to match them - we just need to remove
the `.bam` suffix. 

It is also *critical* to ensure that the samples in the columns are in the same
order as the rows of `sampleinfo`. When we load these objects into DESeq2 for 
the analysis it will not guess which row of the `sampleinfo` belongs to which
column of the counts matrix, it will assume the same order.

We'll use the new commands:

* `column_to_rownames` to set the rownames using a named column
* `rename_all` which allows to rename all the columns using a string function

```{r createCountMatrix}
countdata <- seqdata %>%
    column_to_rownames("Geneid") %>% # turn the geneid column into rownames
    rename_all(str_remove, ".bam") %>% # remove the ".bam" from the column names
    select(sampleinfo$Sample) %>% # keep sample columns using sampleinfo$Sample
    as.matrix()

# check manual for help
# ?tibble::column_to_rownames 
# ?dplyr::select

head(countdata)
```

Here, we used `str_remove` to remove the unwanted suffix from the column names.
The `stringr` package has a lots of useful functions for manipulating strings 
(text), e.g. `str_replace` or `str_extract`.


# Filtering the genes

<!-- prefiltering -->

For many analysis methods it is advisable to filter out as many genes as 
possible before the analysis to decrease the impact of multiple testing
correction on false discovery rates. This is normally done
by filtering out genes with low numbers of reads and thus likely to be 
uninformative.

With `DESeq2` this is however not necessary as it applies `independent
filtering` during the analysis. On the other hand, some filtering for 
genes that are very lowly expressed does reduce the size of the data matrix, 
meaning that less memory is required and processing steps are carried out 
faster.

We will keep all genes where the total number of reads across all samples is 
greater than 5.

```{r filterGenes}
# check dimension of count matrix
dim(countdata)
# for each gene, compute total count and compare to threshold
# keeping outcome in vector of 'logicals' (ie TRUE or FALSE, or NA)
keep <- rowSums(countdata) > 5
# summary of test outcome: number of genes in each class:
table(keep, useNA="always") 
#Â subset genes where test was TRUE
countdata <- countdata[keep,]
# check dimension of new count matrix
dim(countdata)
```

# Quality assessment

Before moving on to doing the actual differential expression analysis it is
important to assess the quality of our data.

## Data transformations and visualization

Differential expression calculations with DESeq2 uses raw read counts as input,
but for visualization purposes we use transformed counts.

### Raw counts 

Why not raw counts? Two issues:

* Raw counts range is very large
* Variance increases with mean gene expression, this has impact on assessing
  the relationships.

```{r raw_summary}
summary(countdata)
```

```{r raw_boxplot}
# few outliers affect distribution visualization
boxplot(countdata, main='Raw counts', las=2)
```

```{r raw_mean_vs_sd}
# Raw counts mean expression Vs standard Deviation (SD)
plot(rowMeans(countdata), rowSds(countdata), 
     main='Raw counts: sd vs mean', 
     xlim=c(0,10000),
     ylim=c(0,5000))
```

<!--
<p>&nbsp;</p>
<p>&nbsp;</p>
-->

### Data transformation

To avoid problems posed by raw counts, they can be [transformed](http://www.bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#data-transformations-and-visualization).
Several transformation methods exist to limit the dependence of variance on mean gene expression:

* Simple log2 transformation
* VST : variance stabilizing transformation
* rlog : regularized log transformation

<p>&nbsp;</p>

#### log2 transformation

Because some genes are not expressed (detected) in some samples, their count are `0`. As log2(0) returns -Inf in R which triggers errors by some functions, we add 1 to every count value to create 'pseudocounts'. The lowest value then is 1, or 0 on the log2 scale (log2(1) = 0).

```{r logTransform}
# Get log2 counts
logcounts <- log2(countdata + 1)
# summary(logcounts[,1]) # summary for first column
# summary(logcounts) # summary for each column
```

We will check the distribution of read counts using a boxplot and add some
colour to see if there is any difference between sample groups.

```{r plotLogCounts}
# make a colour vector
statusCol <- match(sampleinfo$Status, c("virgin", "pregnant", "lactate")) + 1
# '+1' to avoid color '1' i.e. black

# Check distributions of samples using boxplots
boxplot(logcounts,
        xlab="",
        ylab="Log2(Counts)",
        las=2,
        col=statusCol,
        main="Log2(Counts)")
# Let's add a blue horizontal line that corresponds to the median
abline(h=median(logcounts), col="blue")
```

From the boxplots we see that overall the density distributions of raw
log-counts are not identical but still not very different. If a sample is
really far above or below the blue horizontal line (overall median) we may need to investigate
that sample further.

```{r log2_mean_vs_sd}
# Log2 counts standard deviation (sd) vs mean expression
plot(rowMeans(logcounts), rowSds(logcounts), 
     main='Log2 Counts: sd vs mean')
```

In contrast to raw counts, with log2 transformed counts lowly expressed genes show higher variation.

<p>&nbsp;</p>
<p>&nbsp;</p>

#### VST : variance stabilizing transformation

Variance stabilizing transformation (VST) aims at generating a matrix of values for which variance is constant across the range of mean values, especially for low mean.

The `vst` function computes the fitted dispersion-mean relation, derives the transformation to apply and accounts for library size.

```{r vst_counts}
vst_counts <- vst(countdata)

# make a colour vector
statusCol <- match(sampleinfo$Status, c("virgin", "pregnant", "lactate")) + 1

# Check distributions of samples using boxplots
boxplot(vst_counts, 
        xlab="", 
        ylab="VST counts",
        las=2,
        col=statusCol)
# Let's add a blue horizontal line that corresponds to the median
abline(h=median(vst_counts), col="blue")
```

```{r vst_mean_vs_sd}
# VST counts standard deviation (sd) vs mean expression
plot(rowMeans(vst_counts), rowSds(vst_counts), 
     main='VST counts: sd vs mean')
```

<p>&nbsp;</p>
<p>&nbsp;</p>

> ### Challenge 1
>
> 1. Use the `DESeq2` function `rlog` to transform the count data. This function
> also normalises for library size.
> 2. Plot the count distribution boxplots with this data  
>    How has this affected the count distributions?

```{r proProcObjChal1, echo=FALSE, eval=FALSE}
# have object ready for challenge
# need countdata and sampleinfo
save(countdata, sampleinfo, file="Robjects/preprocessingChal1.RData")
```

```{r solutionChallenge1}

```

<p>&nbsp;</p>
<p>&nbsp;</p>

## Principal Component Analysis

A principal component analysis (PCA) is an example of an unsupervised analysis,
where we don't specify the grouping of the samples. If the experiment is well
controlled and has worked well, we should find that replicate samples cluster 
closely, whilst the greatest sources of variation in the data should be between
treatments/sample groups. It is also an incredibly useful tool for checking for 
outliers and batch effects.

To run the PCA we should first normalise our data for library size and transform
to a log scale. DESeq2 provides two separate commands to do this (`vst` and `rlog`). Here we
will use the command `rlog`. `rlog` performs a log2 scale transformation in a 
way that compensates for differences between samples for genes with low read 
count and also normalizes between samples for library size. 

You can read more about `rlog`, its alternative `vst` and the comparison
between the two 
[here](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#count-data-transformations).

To plot the PCA results we will use the `autoplot` function from the
`ggfortify` package [@Tang2016]. `ggfortify` is built on top of `ggplot2` and
is able to recognise common statistical objects such as PCA results or linear
model results and automatically generate summary plot of the results in an
appropriate manner.

```{r pcaPlot, message = FALSE, fig.width=6.5, fig.height=5, fig.align="center"}
library(ggfortify)

rlogcounts <- rlog(countdata)

# run PCA
pcDat <- prcomp(t(rlogcounts))
# plot PCA
autoplot(pcDat)
```

We can use colour and shape to identify the Cell Type and the Status of each sample.

```{r pcaPlotWiColor, message = FALSE, fig.width=6.5, fig.height=5, fig.align="center"}
autoplot(pcDat,
         data = sampleinfo, 
         colour="CellType", 
         shape="Status",
         size=5)
```

> ### Discussion

What does the PCA plot tell us?

<!-- amount and source of variation, sample swap, interaction -->

Let's identify these samples. The package `ggrepel` allows us to add text to 
the plot, but ensures that points that are close together don't have their
labels overlapping (they *repel* each other).

```{r badSamples, fig.width=6.5, fig.height=5, fig.align="center"}
library(ggrepel)

# setting shape to FALSE causes the plot to default to using the labels instead of points
autoplot(pcDat,
         data = sampleinfo, 
         colour="CellType", 
         shape="Status",
         size=5) +
    geom_text_repel(aes(x=PC1, y=PC2, label=Sample), box.padding = 0.8)
```

The mislabelled samples are *MCL1.DG*, which is labelled as *luminal* but should
be *basal*, and *MCL1.LA*, which is labelled as *basal* but should be *luminal*.
Let's fix the sample sheet.

We're going to use another `dplyr` command: 
`mutate`. `mutate` creates new columns in the data frame.

```{r correctSampleSheet}
sampleinfo <- sampleinfo %>% 
    mutate(CellType=ifelse(Sample=="MCL1.DG", "basal", CellType)) %>% 
    mutate(CellType=ifelse(Sample=="MCL1.LA", "luminal", CellType))
```

...and export it so that we have the correct version for later use.

```{r, exportSampleSheet, eval=FALSE}
write_tsv(sampleinfo, "results/SampleInfo_Corrected.txt")
```

Let's look at the PCA now.

```{r correctedPCA, fig.width=6.5, fig.height=5, fig.align="center"}
autoplot(pcDat,
         data = sampleinfo, 
         colour="CellType", 
         shape="Status",
         size=5)
```

Replicate samples from the same group cluster together in the plot, while 
samples from different groups form separate clusters. This indicates that the
differences between groups are larger than those within groups.
The biological signal of interest is stronger than the noise (biological and technical) and can be detected.
Also, differences between virgin, pregnant and lactating are greater for luminal than basal cells.

Clustering in the PCA plot can be used to motivate changes to the design
matrix in light of potential batch effects. For example, imagine that the
first replicate of each group was prepared at a separate time from the second
replicate. If the PCA plot showed separation of samples by time, it might be
worthwhile including time in the downstream analysis to account for the
time-based effect.

## Export data

**We can save a few data objects to use later so we don't have to rerun 
everything**

```{r saveData, eval=F}
save(countdata, sampleinfo, file="results/preprocessing.RData")
```

--------------------

# References
