---
title: "RNA-seq analysis in R"
author: "Stephane Ballereau, Dominique-Laurent Couturier, Mark Dunning, Abbi Edwards, Ashley Sawle"
date: '`r format(Sys.time(), "Last modified: %d %b %Y")`'
output:
  html_document:
    toc: yes
  html_notebook:
    toc: yes
minutes: 300
layout: page
subtitle: Gene Set Testing for RNA-seq
bibliography: ref.bib
editor_options: 
  chunk_output_type: inline
---

```{r setup, include=FALSE, cache=FALSE}
library(tidyverse)
```

The list of differentially expressed genes is sometimes so long that its 
interpretation becomes cumbersome and time consuming. It may also be very
short while some genes have low p-value yet higher than the given threshold.

A common downstream procedure to combine information across genes is gene set testing.
It aims at finding pathways or gene networks the differentially expressed genes play a role in.

Various ways exist to test for enrichment of biological pathways. We will look
into over representation and gene set enrichment analysis.

A gene set comprises genes that share a biological function, chromosomal location, or any other
relevant criterion.

<!--
- Define gene set
- gene set == "pathway"
- over-represented == enriched
ie pathway A is enriched in our diff exp gene list
- 
-->

# Over-representation

## Method

This method tests whether genes in a pathway are present in a subset of our data
more than expected (explanations derived from the [clusterProfiler manual](https://yulab-smu.github.io/clusterProfiler-book/index.html)).

Genes is the experiment are split in two ways:

* annotated to the pathway or not
* differentially expressed or not

Contingency table with:

* rows: in pathway or not
* columns: differentially expressed or not

Example:

```{r}
d <- data.frame(
  diffExpNo=c(1980, 17920),
  diffExpYes=c(20, 80))
row.names(d) <- c("pathwayYes", "pathwayNo")
d
```

<!--
Draws == DEGs
Success == in pathway
-->

For a given pathway:

* N: total number of genes in the background set, e.g. all genes tested
* M: number of genes within that background distribution that are annotated to the pathway
* n: number of differentially expressed genes
* k: number of differentially expressed genes that are annotated to the pathway

Significance can be assessed with a the hypergeometric distribution:

<!-- over-representation p-value -->
![](images/overRepPvalHyperGeoFormula.png)

The test above is identical to the one-tailed Fisher's exact test.

```{r}
fisher.test(d, alternative = "greater")
```


<!--
https://yulab-smu.github.io/clusterProfiler-book/index.html
https://yulab-smu.github.io/clusterProfiler-book/chapter2.html#over-representation-analysis
-->

<!-- KEGG pathway enrichment analysis -->

## `clusterProfiler`

`clusterprofiler` [@Yu2012] supports direct online access of the current KEGG
database, rather than relying on R annotation packages, it also provides some
nice visualisation options (KEGG: Kyoto Encyclopedia of Genes and Genomes).

```{r loadClusterProfiler, message=FALSE}
library(clusterProfiler)
search_kegg_organism('mmu', by='kegg_code')
```

## KEGG enrichment analysis

The input for the KEGG enrichment is list of gene IDs for significant genes. 

We now load the R object keeping the outcome of the differential expression 
analysis for the LvV contrast.

```{r loadData}
load("Robjects/Annotated_Results_LvV.RData")
```

We will only use genes with an absolute fold change greater than 2.

For this tool we need to use Entrez IDs, so we will eliminate genes with no such ID
by filtering on missing values in 'Entrez'.

```{r enrichKEGG}
sigGenes <- shrinkLvV %>% 
    filter(FDR < 0.05 & !is.na(FDR) & 
               abs(logFC) > 1 & 
               !is.na(Entrez)) %>% 
    pull(Entrez)

kk <- enrichKEGG(gene = sigGenes, organism = 'mmu')
head(kk, n=10)
head(kk, n=10) %>% as_tibble()
```

## Visualise a pathway

### In a browser

`clusterProfile` has a function `browseKegg` that allows you to view the KEGG
pathway with the genes colours in in your browser.

```{r browseKegg}
browseKEGG(kk, 'mmu03320')
```

### As a file

The package `pathview` [@Luo2013] can be used to generate figures of KEGG 
pathways. 

One advantage over `clusterProfiles` browser method is that the genes are 
coloured according to their fold change levels in our data. To do this we need
to pass `pathview` a named vector of fold change data (actually you could
colour by any numeric vector, e.g. p-value).

The package plots the KEGG pathway to a `png` file in the working directory.

```{r pathview, message=F}
library(pathview)
logFC <- annotLvV$logFC
names(logFC) <- annotLvV$Entrez
pathview(gene.data = logFC, 
         pathway.id = "mmu03320", 
         species = "mmu", 
         limit = list(gene=5, cpd=1))
```

mmu03320.pathview.png:

![mmu03320 - PPAR signaling pathway](results/mmu03320.pathview.png)

> ## Challenge 1 {.challenge}
>
> 1. Use `pathview` to export a figure for "mmu04060", but this time only
> use genes that are statistically significant at FDR < 0.01

```{r solution1, eval=F}

```

# GSEA analysis

Gene Set Enrichment Analysis (GSEA) identifies gene sets that are related to the
difference of interest between samples [@Subramanian15545].

The software is distributed by the [Broad Institute](http://software.broadinstitute.org/gsea/index.jsp)
and is freely available for use by academic and non-profit organisations.
The Broad also provide a number of very well curated gene sets
for testing against your data - the [Molecular Signatures Database (MSigDB)](http://software.broadinstitute.org/gsea/msigdb/index.jsp). 
These are collections of human genes, however. Fortunately, these lists
have been translated to mouse equivalents by the Walter+Eliza Hall Institute
Bioinformatics service and made available for [download](http://bioinf.wehi.edu.au/software/MSigDB/).

## Concept

We will look into random walk, distributions and the Kolomogorov-Smirnoff test
whose concept GSEA uses (inspired from [this course](https://www.coursera.org/lecture/network-biology/gene-set-enrichment-analysis-gsea-preliminaries-C4GxN)).

### Random walks

A random walk is a succession of random steps along a mathematical path.
of the distance from 0.

```{r}
# a path 
plot.new()
plot.window(xlim=c(-10,10), ylim=c(-5,5))
axis(side=1)
```

Starting at 0, we will take a step randomly in either direction and keep track
of the distance from 0.

A short walk:

```{r}
# random walk
rn <- rnorm(n=10,m=0,sd=0.5)
rwalk <- cumsum(sign(rn))
plot(rwalk, type="l", ylim=c(-5,5))
abline(h=0)
```

A longer walk:

```{r}
rn <- rnorm(n=1000,m=0,sd=0.5)
rwalk <- cumsum(sign(rn))
plot(rwalk, type="l")
abline(h=0)
```

We can force the walk to end at 0, by having the same number of steps in both
directions.

```{r}
# draw observations
rn <- rnorm(n=1000,m=0,sd=0.5)
# find indices of positive values (steps to the right)
whichPos <- which(rn >= 0)
# find indices of negative values (steps to the left)
whichNeg <- which(rn < 0)
# check number of steps in each direction
length(whichPos)
length(whichNeg)
# concat indices for the same number of steps in each direction, eg 400 steps each
rn2 <- c(rn[whichNeg][1:400], rn[whichPos][1:400])
# re-shuffle
rn2 <- sample(rn2)
# compute walk
rwalk <- cumsum(sign(rn2))
# plot
plot(rwalk, type="l")
abline(h=0)
```

The probability of observing the maximum distance observed here can be computed.

### Distributions

A brief description of probability distributions. We will use a example of a
variable following a Normal distribution of mean 0 and standard deviation 0.5.

```{r}
par(mfrow = c(1, 2))
    
# probability density function

xx <- seq(-2, 2, length=100)
plot(xx, dnorm(xx, mean=0, sd=0.5), type='l')

# cumulative distribution function

cumulative <- pnorm(xx, 0, 1)
plot(xx, cumulative, type='l')
```

### Kolmogorov-Smirnov test

The Kolmogorov-Smirnov (KS) test of goodness-of-fit tests whether the observed data is consistent
with a given cumulative density function (CDF).

To illustrate the method, we will keep the random variable following the Normal distribution. We will make some observations and plot both the observed and expected CDFs.

Because the observed data was obtained with a finite sample, the observed CDF will differ from the theoretical CDF by some random scatter.

The KS test considers that if there is no difference between the two CDFs then that difference should be a
random walk, with both ends fixed at 0. For which we can compute the probability
of observing a given distance.

Example:

```{r}
rn <- rnorm(n=20, m=0, sd=0.5)
X <- rn
X <- sort(X)
e_cdf <- 1:length(X) / length(X)
plot(X, e_cdf, type = "s", xlim=c(-1,1), ylim=c(0,1))

xx <- seq(min(X), max(X), length=20)
cumulative <- pnorm(xx, 0, 0.5)
lines(xx, cumulative)
```

Another example, where the two distributions differ. The random variable now
follows a Normal distribution of mean -0.5 (and standard deviation still at
0.5).

```{r}
rn <- rnorm(n=20, m=-0.3, sd=0.5)
X <- rn
X <- sort(X)
e_cdf <- 1:length(X) / length(X)
plot(X, e_cdf, type = "s", xlim=c(-1,1), ylim=c(0,1))

xx <- seq(min(X), max(X), length=20)
cumulative <- pnorm(xx, 0, 0.5)
lines(xx, cumulative)
```

## GSEA

The analysis is performed by:

- (i) ranking all genes in the data set
- (ii) identifying the rank positions of all members of the gene set in the 
ranked data set
- (iii) calculating an enrichment score (ES) that represents the difference 
between the observed rankings and that which would be expected assuming a random 
rank distribution.
 
The article describing the original software is available 
[here](http://www.pnas.org/content/102/43/15545.long), while this [commentary on GSEA](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1266131/) provides a shorter description.

![](images/gseaArticleFig1.png)

We will use the [`fgsea`](http://www.bioconductor.org/packages/release/bioc/vignettes/fgsea/inst/doc/fgsea-tutorial.html) 
package [@Sergushichev2016] that implements the same algorithm in R. 'fgsea' stands for' "fast preranked gene set enrichment analysis (GSEA)".

```{r fgsea}
library(fgsea)
```

## Ranking Data

We need to provide `fgsea` a vector containing numeric data by which it should 
rank the genes. To start with we will simply use a rank based on their fold 
change. We do not need to rank the genes ourselves, `fgsea` will do this based
on the data we provide.

We must exclude genes for which we do not have Entrez IDs. Also, we should 
use the shrunk LFC values.

```{r preparedata}
gseaDat <- filter(shrinkLvV, !is.na(Entrez))

rankData <- gseaDat$logFC
names(rankData) <- gseaDat$Entrez
head(rankData)
```

## Load pathways

```{r loadPathways}
load("Robjects/mouse_H_v5.RData")
pathwaysH <- Mm.H
```

## Conduct analysis

```{r runFgsea}
fgseaRes <- fgsea(pathwaysH, 
                  rankData, 
                  minSize = 15, 
                  maxSize = 500, 
                  nperm = 1000)
```

The warning produced indicates that there are few genes that have the same fold
change and so are ranked equally. `fgsea` arbitrarily determines which
comes first in the ranked list. As long as this number is small it shouldn't
significantly affect the results. If the number is large something is 
suspicious about the fold change results.

Lets look at the top 10 results.

```{r top10pathways, eval=FALSE}
fgseaRes %>% 
    arrange(desc(abs(NES))) %>% 
    top_n(10, -padj)
```

```{r top10pathwaysPrint, echo=FALSE}
fgseaRes %>% 
    arrange(desc(abs(NES))) %>% 
    top_n(10, -padj) %>%  
    as_tibble()
```

## Enrichment score plot

```{r gseaEnrichmentPlot}
plotEnrichment(pathwaysH[["HALLMARK_ESTROGEN_RESPONSE_EARLY"]], rankData)
```

Remember to check the [GSEA 
article](http://www.pnas.org/content/102/43/15545.full) for the complete 
explanation.

> ## Challenge 2 {.challenge}
>
> Another common way to rank the genes is to order by pvalue, but also, sorting
> so that upregulated genes are at start and downregulated at the other - 
> you can do this combining the sign of the fold change and the pvalue.  
> 1. Rank the genes by statisical significance - you will need to create
> a new ranking value using `-log10({p value}) * sign({Fold Change})`  
> 2. Load the "C2" pathways from the the `Robjects/mouse_c2_v5.RData` file   
> 3. Run `fgsea` using the new ranked genes and the C2 pathways  
> 4. Run `fgsea` using the new ranked genes and the H pathways. How do these 
> results differ from the ones we got when ranking by the fold change alone?  

```{r solution2}

```

---------------------------------------------------------------

# References
